tests/integration/test_upgrade_charm.py

```diff
@@ -28,12 +52,77 @@ async def test_deploy_from_edge_and_upgrade_from_local_path(ops_test, prometheus
-        await ops_test.model.deploy(f"ch:{app_name}", application_name=app_name, channel="edge")
-        await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=TIMEOUT)
+        await ops_test.model.deploy(
+            f"ch:{PROMETHEUS_APP_NAME}", application_name=PROMETHEUS_APP_NAME, channel="edge"
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[PROMETHEUS_APP_NAME], status="active", timeout=TIMEOUT
+        )
 
         logger.debug("upgrade deployed charm with local charm %s", prometheus_charm)
-        await ops_test.model.applications[app_name].refresh(
-            path=prometheus_charm, resources=resources
+        await ops_test.model.applications[PROMETHEUS_APP_NAME].refresh(
+            path=prometheus_charm, resources=prometheus_resources
         )
-        await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=TIMEOUT)
-        await check_prometheus_is_ready(ops_test, app_name, 0)
+        await ops_test.model.wait_for_idle(
+            apps=[PROMETHEUS_APP_NAME], status="active", timeout=TIMEOUT
+        )
+        await check_prometheus_is_ready(ops_test, PROMETHEUS_APP_NAME, 0)
+
+
+async def test_metrics_endpoint_provider_can_be_upgraded(ops_test, prometheus_tester_charm):
+    """Ensure scrape alert rules can be updated.
+
+    This test upgrades the metrics provider charm and checks that
+    updates to alert rules are propagated correctly.
+    """
+    await ops_test.model.deploy(
+        prometheus_tester_charm, resources=tester_resources, application_name=TESTER_APP_NAME
+    )
+    await ops_test.model.wait_for_idle(apps=[TESTER_APP_NAME], status="active")
+    await ops_test.model.block_until(
+        lambda: len(ops_test.model.applications[TESTER_APP_NAME].units) > 0
+    )
+    assert ops_test.model.applications[TESTER_APP_NAME].units[0].workload_status == "active"
+
+    await ops_test.model.add_relation(PROMETHEUS_APP_NAME, TESTER_APP_NAME)
+    await ops_test.model.wait_for_idle(
+        apps=[PROMETHEUS_APP_NAME, TESTER_APP_NAME], status="active"
+    )
+
+    # Check only one alert rule exists
+    tester_rules = []
+    for attempt in Retrying(
+        retry=retry_if_exception_type(AssertionError), stop=stop_after_attempt(3)
+    ):
+        try:
+            with attempt:
+                rules_with_relation = await get_prometheus_rules(ops_test, PROMETHEUS_APP_NAME, 0)
+                tester_rules = get_rules_for(TESTER_APP_NAME, rules_with_relation)
+                assert len(tester_rules) == 1
+        except RetryError:
+            pass
+
+    # Add new alert rule, rebuild and refresh prometheus tester charm
+    write_tester_alert_rule_file(MISSING_TARGET_RULE, "target_missing.rule")
+    tester_charm = await rebuild_prometheus_tester(ops_test)
+    await ops_test.model.applications[TESTER_APP_NAME].refresh(
+        path=tester_charm, resources=tester_resources
+    )
+    remove_tester_alert_rule_file("target_missing.rule")
```

Comment:
Since the extra rule is needed only for packing, maybe create a context manager for it?
```python
with write_tester_alert_rules(...) as tester_charm:
    # ...
```

Reply:
I find it simpler to just provide explicit methods to write and remove the file instead of an "enter" and "exit" method. A context manager would be useful if we need to write multiple files in different contexts, for which I do not see a use case at present. Also I find the phrase "with write_tester_alert_rules() as tester_charm" odd to read.

Reply:
Resolving as discussed.